### PR TITLE
box: forbid non-string types in key_def.new()

### LINKED ISSUE #5222: key_def: incorrect validation for part type
Tarantool version: 2.3

```lua
tarantool> key_def.new({{field = 1, type = nil}})
---
- error: type must not be nil
...
tarantool> key_def.new({{field = 1, type = box.NULL}}) -- Expected the same error as for first case
---
- - type: unsigned
    is_nullable: false
    fieldno: 1
...
```

### LINKED ISSUE #5940: box: a partial 'num' and 'str' field type should not be accepted
## How to reproduce

```yaml
tarantool> box.cfg{}
tarantool> box.schema.create_space('a', {format = {[1] = {name = 'id', type = ''}}})
tarantool> box.space.a:format()
---
- [{'name': 'id', 'type': ''}]
...
tarantool> box.schema.create_space('b', {format = {[1] = {name = 'id', type = 'st'}}})
tarantool> box.space.b:format()
---
- [{'name': 'id', 'type': 'st'}]
...
```

## Solution

```diff
commit 3657390813d7b5d55d268db1021f6664d49c5a58
Author: Alexander Turenko <alexander.turenko@tarantool.org>
Date:   Fri Sep 4 01:27:17 2020 +0300

    WIP: box: strengthen field name check
    
    Don't accept an empty string or leading part of "str" or "num" as a
    valid field type.
    
    XXX: add tests (using key_def lua module?)

diff --git a/src/box/field_def.c b/src/box/field_def.c
index 213e91699..e90a389cb 100644
--- a/src/box/field_def.c
+++ b/src/box/field_def.c
@@ -181,9 +181,9 @@ field_type_by_name(const char *name, size_t len)
        if (field_type != field_type_MAX)
                return field_type;
        /* 'num' and 'str' in _index are deprecated since Tarantool 1.7 */
-       if (strncasecmp(name, "num", len) == 0)
+       if (len == 3 && strncasecmp(name, "num", len) == 0)
                return FIELD_TYPE_UNSIGNED;
-       else if (strncasecmp(name, "str", len) == 0)
+       else if (len == 3 && strncasecmp(name, "str", len) == 0)
                return FIELD_TYPE_STRING;
        else if (len == 1 && name[0] == '*')
                return FIELD_TYPE_ANY;
```

### PR DESCRIPTION
**Pull request by Gumix**
Currently `lua_tolstring` in `luaT_key_def_set_part` returns `type_len == 0`, then `strncasecmp(name, "num", 0)` in `field_type_by_name` returns 0, which results in `FIELD_TYPE_UNSIGNED` instead of an error.

Closes #5222
Closes #5940

### CODE REVIEW

File: src/box/lua/key_def.c [hunk 1]
```diff
@@ -124,12 +124,15 @@ luaT_key_def_set_part(struct lua_State *L, struct key_part_def *part,
 	/* Set part->type. */
 	lua_pushstring(L, "type");
 	lua_gettable(L, -2);
-	if (lua_isnil(L, -1)) {
+	size_t type_len;
+	const char *type_name;
+	bool type_is_nil = lua_isnil(L, -1);
+	if (!type_is_nil)
+		type_name = lua_tolstring(L, -1, &type_len);
+	if (type_is_nil || type_len == 0) {
```
Comment (locker) [inline src/box/lua/key_def.c:132]: This results in confusing behavior:

```
tarantool> require('key_def').new({{field = 1, type = ''}})
---
- error: type must not be nil
...

tarantool> require('key_def').new({{field = 1, type = true}})
---
- error: type must not be nil
...
```

I think that we should explicitly check that the type is a string here and fix the error message.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/lua/key_def.c (hunk 1 above)
Comment (Totktonada) [reply src/box/lua/key_def.c]: I suggest to use `lua_typename(L, lua_type(L, idx))` to generate an error message with the actual Lua type. Example:

https://github.com/tarantool/tarantool/blob/6e45042473fb973c7b1ad87fcf04c9081a4c40fa/src/lua/popen.c#L618-L631

Comment (Gumix) [reply src/box/lua/key_def.c]: Done

File: src/box/lua/key_def.c [hunk 2]
```diff
@@ -124,15 +124,15 @@ luaT_key_def_set_part(struct lua_State *L, struct key_part_def *part,
...
 	}
 	size_t type_len;
 	const char *type_name = lua_tolstring(L, -1, &type_len);
 	lua_pop(L, 1);
 	part->type = field_type_by_name(type_name, type_len);
-	if (part->type == field_type_MAX) {
+	if (type_len == 0 || part->type == field_type_MAX) {
 		diag_set(IllegalParams, "Unknown field type: %s", type_name);
 		return -1;
 	}
```
[the remark this answers is not in the document]
Comment (Gumix) [reply src/box/lua/key_def.c]: Done

File: changelogs/unreleased/gh-5940-strengthen-field-type-check.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+## feature/box
+
+* **[Breaking change]** Now the empty string, 'n', 'nu', 's' and 'st' are
+  not accepted as valid field types (gh-5940).
```
Comment (Totktonada) [inline changelogs/unreleased/gh-5940-strengthen-field-type-check.md:4]: And '' (an empty string).

Comment (Gumix) [reply changelogs/unreleased/gh-5940-strengthen-field-type-check.md:4]: «… Now the empty string …»

Comment (Totktonada) [reply changelogs/unreleased/gh-5940-strengthen-field-type-check.md:4]: Ouch, sorry.

**Review by Totktonada — APPROVED** [review]

File: changelogs/unreleased/gh-5940-strengthen-field-type-check.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+## feature/box
```
Comment (locker) [inline changelogs/unreleased/gh-5940-strengthen-field-type-check.md:1]: Let's change to bugfix.

**Review by locker — CHANGES_REQUESTED** [review]

**Comment by locker** [comment]
Cherry-picked to 2.10.

**Pull request merged**